### ISSUE TITLE
external kubernetes: proxied API calls are extremely slow compared to base Kube

### ISSUE THREAD

**Issue by paralin** [openshift/origin#8215]
When I do a call to the base kube cluster:
```
$ time kubectl get pods --context=aws_k8s
NAME                    READY     STATUS    RESTARTS   AGE
docker-registry-jky2n   1/1       Running   0          11h
kubectl get pods --context=aws_k8s  0.17s user 0.00s system 21% cpu 0.824 total
```

Timing: 0.82 seconds.

When I do a call to the openshift cluster:

```
$ time kubectl get pods --namespace=default
NAME                    READY     STATUS    RESTARTS   AGE
docker-registry-jky2n   1/1       Running   0          11h
kubectl get pods --namespace=default  0.03s user 0.00s system 0% cpu 10.869 total
```

Timing: 10.86 seconds! Why! That's a ridiculous amount of time for such a simple call.

With a higher log level I see that the initial call takes almost the entire time:

```
I0323 10:37:45.947695   12460 round_trippers.go:267] curl -k -v -XGET  -H "Accept: application/json, */*" -H "Authorization: Bearer mytokenhere" https://os.mydomain.com:443/api
```

That's around 9 seconds while the rest of the calls are 1 second ish. Why is this? Is it related to #2304?

**Comment by liggitt** [comment]
not related to that issue. the same call for me takes ~100ms. if you make the same call again, is it faster? wondering if it is a DNS issue between openshift and the backend kube

**Comment by paralin** [comment]
@liggitt nope it's the same timing always. quite frustrating.

**Comment by liggitt** [comment]
can you provide the output of the following:
```
curl -k https://os.mydomain.com:443/api
curl -k https://<basekube>:443/api
```

**Comment by paralin** [comment]
@liggitt 

```
curl -k https://os.mydomain.com:443/api -v
*   Trying 54.194.100.27...
* Connected to os.paral.in (54.194.100.27) port 443 (#0)
* found 173 certificates in /etc/ssl/certs/ca-certificates.crt
* found 704 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
*        common name: 10.0.182.162 (matched)
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: CN=10.0.182.162
*        start date: Wed, 23 Mar 2016 03:11:13 GMT
*        expire date: Fri, 23 Mar 2018 03:11:14 GMT
*        issuer: CN=openshift-signer@1458702672
*        compression: NULL
* ALPN, server accepted to use http/1.1
> GET /api HTTP/1.1
> Host: os.paral.in
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Cache-Control: no-store
< Location: /api/
< Date: Wed, 23 Mar 2016 15:00:50 GMT
< Content-Length: 40
< Content-Type: text/html; charset=utf-8
< 
<a href="/api/">Moved Permanently</a>.

* Connection #0 to host os.mydomain.com left intact
```

and the other in a sec...

**Comment by liggitt** [comment]
Sorry, with the trailing /
Interested in the content

**Comment by paralin** [comment]
Going to forget about redacting these at this point, I'm about to nuke this cluster anyway...

```
curl -k https://os.paral.in:443/api/ -v
*   Trying 54.194.100.27...
* Connected to os.paral.in (54.194.100.27) port 443 (#0)
* found 173 certificates in /etc/ssl/certs/ca-certificates.crt
* found 704 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
*        common name: 10.0.182.162 (matched)
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: CN=10.0.182.162
*        start date: Wed, 23 Mar 2016 03:11:13 GMT
*        expire date: Fri, 23 Mar 2018 03:11:14 GMT
*        issuer: CN=openshift-signer@1458702672
*        compression: NULL
* ALPN, server accepted to use http/1.1
> GET /api/ HTTP/1.1
> Host: os.paral.in
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Cache-Control: no-store
< Content-Length: 181
< Content-Type: application/json
< Date: Wed, 23 Mar 2016 15:10:26 GMT
< 
{
  "kind": "APIVersions",
  "versions": [
    "v1"
  ],
  "serverAddressByClientCIDRs": [
    {
      "clientCIDR": "0.0.0.0/0",
      "serverAddress": "172.20.0.9:443"
    }
  ]
* Connection #0 to host os.paral.in left intact
}
```

**Comment by paralin** [comment]
And to the base kube...

```
curl -k https://52.50.85.106/api/ -v   
*   Trying 52.50.85.106...
* Connected to 52.50.85.106 (52.50.85.106) port 443 (#0)
* found 173 certificates in /etc/ssl/certs/ca-certificates.crt
* found 704 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
*        common name: k8s-master (matched)
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: CN=k8s-master
*        start date: Wed, 23 Mar 2016 02:51:42 GMT
*        expire date: Sat, 21 Mar 2026 02:51:42 GMT
*        issuer: CN=52.50.85.106@1458701502
*        compression: NULL
* ALPN, server accepted to use http/1.1
> GET /api/ HTTP/1.1
> Host: 52.50.85.106
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 401 Unauthorized
< Content-Type: text/plain; charset=utf-8
< Www-Authenticate: Basic realm="kubernetes-master"
< Date: Wed, 23 Mar 2016 15:11:15 GMT
< Content-Length: 13
< 
Unauthorized
* Connection #0 to host 52.50.85.106 left intact

```

**Comment by paralin** [comment]
And now confirming to base kube it's 12 milliseconds. What.

**Comment by paralin** [comment]
Hmm weird @liggitt it's really fast now, with a fresh cluster.

**Comment by paralin** [comment]
@liggitt Slow again... Here's what's happening:

```
I0325 14:16:56.391180   24762 loader.go:242] Config loaded from file /home/paralin/.kube/config
I0325 14:16:56.391954   24762 round_trippers.go:264] GET https://openshift.mydomain.com:443/oapi
I0325 14:16:56.391974   24762 round_trippers.go:271] Request Headers:
I0325 14:16:56.391986   24762 round_trippers.go:274]     Authorization: Bearer mytoken
I0325 14:16:56.391999   24762 round_trippers.go:274]     User-Agent: oc/v1.1.4 (linux/amd64) openshift/3941102
I0325 14:16:56.392010   24762 round_trippers.go:274]     Accept: application/json, */*
I0325 14:17:01.871326   24762 round_trippers.go:289] Response Status: 200 OK in 5479 milliseconds
I0325 14:17:01.871362   24762 round_trippers.go:292] Response Headers:
I0325 14:17:01.871374   24762 round_trippers.go:295]     Cache-Control: no-store
I0325 14:17:01.871385   24762 round_trippers.go:295]     Content-Type: application/json
I0325 14:17:01.871396   24762 round_trippers.go:295]     Date: Fri, 25 Mar 2016 18:17:01 GMT
I0325 14:17:01.871406   24762 round_trippers.go:295]     Content-Length: 59
I0325 14:17:01.872086   24762 request.go:875] Response Body: {"kind":"APIVersions","apiVersion":"v1","versions":["v1"]}
```

**Comment by liggitt** [comment]
Same questions. What does curl do when you run the same request through it? Are subsequent requests faster? That endpoint is not proxied, so should be unrelated to external kube (and is ~1ms on all systems I see)

**Comment by liggitt** [comment]
If you run the same command again, is it fast, or is the oapi call always slow? Still wondering if DNS resolution is an issue.

**Comment by paralin** [comment]
@liggitt The first call is always slow, even if I run it in quick succession multiple times. 

Might be dns resolution, let me try again with curl

**Comment by paralin** [comment]
@liggitt Curl doesn't output any logging in max verbosity mode for around 6 seconds and then all of the actual calls against OpenShift are fast so I'm thinking maybe you're right about it being DNS resolution slow

**Comment by paralin** [comment]
@liggitt But if I use "dig" to check the DNS it responds within half a second or so.

**Comment by liggitt** [comment]
if curl is seeing the same issue, it seems like a network issue between the openshift and kube machines
